### PR TITLE
PR for Duplicated UDP responses from DNS confuse LookupClient #140

### PR DESCRIPTION
**Pull request by JamesKovacs**
The following PR is to resolve the issue reported in #140.

### CODE REVIEW

**Comment by MichaCo** [comment]
Hi @JamesKovacs 
I think the implementation itself looks reasonable (will review it a bit more now)

The unit test, especially `Lookup_XidMismatch` seems a bit flaky though: The 2 tests with `mismatchResponses` = 10 fail for me locally all the time.

File: test/DnsClient.Tests/LookupTest.cs [hunk 1]
```diff
@@ -1113,5 +1112,73 @@ public void Lookup_SettingsFallback_KeepProvidedServers3()
...
+
+        [Theory]
+        [InlineData(0, true)]
+        [InlineData(1, true)]
+        [InlineData(3, true)]
+        [InlineData(10, true)]
+        [InlineData(0, false)]
+        [InlineData(1, false)]
+        [InlineData(3, false)]
+        [InlineData(10, false)]
```
Comment (MichaCo) [inline test/DnsClient.Tests/LookupTest.cs:1124]:
```suggestion
        //[InlineData(10, false)]
```

File: test/DnsClient.Tests/LookupTest.cs [hunk 2]
```diff
@@ -1113,5 +1112,73 @@ public void Lookup_SettingsFallback_KeepProvidedServers3()
             Assert.NotEqual(client.Settings, settings);
             Assert.NotEqual(client.NameServers, settings.NameServers);
         }
+
+        [Theory]
+        [InlineData(0, true)]
+        [InlineData(1, true)]
+        [InlineData(3, true)]
+        [InlineData(10, true)]
```
Comment (MichaCo) [inline test/DnsClient.Tests/LookupTest.cs:1120]:
```suggestion
        //[InlineData(10, true)]
```

Not sure why those 2 variants fail, at least when run under .NET47x. 
Let's just comment it out for now?

File: src/DnsClient/DnsMessageHandler.cs [hunk 3]
```diff
@@ -48,6 +48,14 @@ internal abstract class DnsMessageHandler
             return false;
         }
 
+        protected static void ValidateResponse(DnsRequestMessage request, DnsResponseMessage response)
+        {
+            if (request.Header.Id != response.Header.Id)
```
Comment (MichaCo) [inline src/DnsClient/DnsMessageHandler.cs:53]: right now, response can be potentially be null at this point, at least if the TCP handler is used.
I'll change that in another PR.
But maybe better do a null check and return right away, if response or request are null.

**Review by MichaCo — CHANGES_REQUESTED** [review]
Looks good overall. Just 2 minor things

**Comment by JamesKovacs** [comment]
Thanks for the feedback. I've addressed your comments and will push another commit to the branch shortly.

**Comment by JamesKovacs** [comment]
Changes pushed and ready for review. Thanks again!

**Review by MichaCo — APPROVED** [review]

**Comment by MichaCo** [comment]
The tests failed during build. I'll fix that later, seems not really related to your changes.
I have to finish one more change and then planning to release this end of the week I think.

Thanks again for your help!

**Comment by djj0809** [comment]
@MichaCo Looking forward to the new version. 

We have been seeing DNS resolution failure in one of our cluster from time to time but was not able to identity the root cause, thanks @JamesKovacs !

**Pull request merged**